swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/work/WorkQueueManager.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.swabbie.work
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.config.Schedule
+import com.netflix.spinnaker.config.SwabbieProperties
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import com.netflix.spinnaker.kork.eureka.RemoteStatusChangedEvent
+import com.netflix.spinnaker.swabbie.CacheStatus
+import com.netflix.spinnaker.swabbie.discovery.DiscoveryActivated
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import org.springframework.stereotype.Component
+import java.lang.IllegalStateException
+import java.time.Clock
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.LocalTime
+import java.time.Duration
+
+/**
+ * Serves as the producer of work for resource handlers, monitors and refills the work queue once all work is complete
+ * @property WorkQueue the queue to contain work for processing by resource handlers
+ */
+@Component
+class WorkQueueManager(
+  private val dynamicConfigService: DynamicConfigService,
+  private val swabbieProperties: SwabbieProperties,
+  private val queue: WorkQueue,
+  private val clock: Clock,
+  private val registry: Registry,
+  private val cacheStatus: CacheStatus
+) : DiscoveryActivated {
+  private val log: Logger = LoggerFactory.getLogger(javaClass)
+  private val queueId = registry.createId("swabbie.redis.queue")
+
+  override val onDiscoveryUpCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = {
+      ifEnabled {
+        ensureLoadedCaches()
+        queue.refillOnEmpty()
+      }
+    }
+
+  override val onDiscoveryDownCallback: (event: RemoteStatusChangedEvent) -> Unit
+    get() = {
+      !ifEnabled {
```

Comment:
Shouldn't this always clear the queue? Because, if an instance goes down in discovery we want to make sure it doesn't do more work (even if it's during a scheduled work time)

Reply:
If we were running one instance yes. But since the queue is shared, we should probably not empty it since other instances UP might be using it.

On a second thought, perhaps we don't even need a redisQueue at all. An internal InMemoryQueue would work fine, especially with the lockManager. That way each instance can manage their in memory queue. I actually like this better.

Reply:
Scratch that. We need it lol

Reply:
And we should only clear the queue if disabled. So this should stay as is aside for the style update